### PR TITLE
Add a callback to customize the value before LRU cache

### LINKED ISSUE #7: [feature] add extra callback layer
imo an extra callback should be in between the LRU cache and SHM cache. Often you'd want the scalar values in the SHM to be converted in other structures (eg. an array/enum into a hash table) in Lua land.

If you add that in a callback it runs once, if not you'd have to test at runtime and run that test over and over again.

### PR DESCRIPTION
**Pull request by jdesgats**
This PR fixes #7: it adds a `lru_callback` option to the mlcache constructor in order to be able to store arbitrary values (we use it to cache functions and cdata objects mostly).

I am open to discuss about the API: I added it in the constructor rather than in the `get` method to not bloat it too much, but it could also go in the `opts` parameter.

In practice, we tend to store always the same kind of data in the same mlcache instance, so this API works for my use case while keeping the `get` method reasonably simple, but I would be happy to hear about other possible workflows.

### CODE REVIEW

**Comment by thibaultcha** [comment]
Hi @jdesgats,

Awesome! Thanks for working on this. About your question as to where the argument belongs, I was more envisioning it belonging to the `get()` method, as part of one of its accepted options. The reason being that we would allow users to granularly chose the serialization format for their cached values:
```lua
-- cache some scalar value, like a number
local v1, err = cache:get("some_number", nil, function() return 123 end)

-- use the same mlcache to cache a table/cdata/etc...
local v1, err = cache:get("some_table", { 
  lru_callback = function(t)
    return t.k1 .. t.k2 -- use a concat instead of the default, expensive JSON serialization
  end
}, function()
    return { k1 = "foo", k2 = "bar" }
end)
```

I agree this further complicates the `get()` method, but the added flexibility makes me think it is worth it. Of course, in real production usage, one would reuse the `opts` and `cb` arguments of each `get()` call, instead of hurting the JIT and the GC like with my above example...

Does that makes sense?

As for more minor concerns, initial thought I had for this API:
1. It should be called with `pcall`, as per your original patch and return caught errors 👍 
2. We should accept 2 return values `value + err`, so that users can report errors by returning `nil + err` (just like the `callback` argument, although support is not complete yet, only forward-compatible). This is because the `pcall` is mostly there as a safe-guard against programming errors or serialization and networking APIs that might return errors, but "soft" errors should be reported as usual by users in the returned values.
3. Would you please stick to using double quotes so that I don't have to update them manually before merging your patch, thank you!

File: lib/resty/mlcache.lua [hunk 1]
```diff
@@ -513,7 +529,12 @@ function _M:get(key, opts, cb, ...)
 
     -- set our own worker's LRU cache
 
-    set_lru(self, key, data, ttl, neg_ttl)
+    data, err = set_lru(self, key, data, ttl, neg_ttl)
+    if data == CACHE_MISS_SENTINEL_LRU then
```
Comment (thibaultcha) [inline lib/resty/mlcache.lua]: Maybe we cab simply ignore the first return value of `set_lru` here? Except when it is `nil`, which indicates an error returned by the user (`nil + err`). If you'd rather test the second return value for that, that's fine too imho.

**Comment by thibaultcha** [comment]
Oh, and feel free to not include support for "soft errors" (`return nil, err`) in the LRU callback right now if you think it is out of scope from what you had in mind. Just ensure forward compatibility (just like `callback` at least :) Thank you!

**Comment by Tieske** [comment]
My 2 cts, fwiw:

1. I don't like `pcall`, even with LuaJIT, I've seen the dreaded "yield across C boundary" error. And by then there usually is no way back. So imo better to be clear that the callbacks run unprotected and errors need to be dealt with by the callbacks.
2. for the callback api, I'd do both, provide a callback to the initializer/opts, and allow an override through `get`.

I'm aware that both are very opinionated, so feel free to ignore 😄

File: README.md [hunk 2]
```diff
@@ -218,6 +218,11 @@ holding the desired options for this instance. The possible options are:
   be used as a pub/sub backend for invalidation events propagation across
   workers. Several mlcache instances can use the same `ipc_shm` (events will
   be namespaced).
+- `lru_callback`: function used to transform the values before setting them
+  in the LRU cache (L1). It might bev useful to store arbitrary Lua objects
```
Comment (Tieske) [inline README.md]: bev -> be

**Comment by thibaultcha** [comment]
Hmm yeah, on second thought, my comment's example makes more sense for an shm-level callback (to avoid the default encoding) than an LRU one.

@Tieske 

> 1. I don't like pcall, even with LuaJIT, I've seen the dreaded "yield across C boundary" error. And by then there usually is no way back. So imo better to be clear that the callbacks run unprotected and errors need to be dealt with by the callbacks.

LuaJIT's resumable VM should take care of this?

> 2. for the callback api, I'd do both, provide a callback to the initializer/opts, and allow an override through get.

Sounds good to me - initially afraid this would complicate the library too much, but overall, it stays simple for simple use-cases, but scales up well with users needs if necessary.

---

Another thought crossed my mind: we should maybe avoid referring to those as "callbacks", and use `l1_serializer` instead, to avoid confusion with the `get()` callback argument and the different role those 2 play. Using `l1_` also means that when implementing the shm serializer, we'll refer to it as `l2_serializer`. What are your thoughts on this?

---

Overall, I would tend to follow this approach for completing this PR:
1. rename this argument to `l1_serializer` unless someone has a counter-argument
2. add `l1_serializer` at the `get()` level (and keeping it as an instantiation argument too)
2. add forward-compatibility support for returning errors from it (this means mostly not checking for the 1st return value to be `nil` whatsoever I think?)

Later on, we can implement the `l2_serializer` to provide different shm-encoding mechanisms depending on the cached value type (there are other, higher priority items imho before we do this anyway).

File: lib/resty/mlcache.lua (hunk 1 above)
Comment (jdesgats) [reply lib/resty/mlcache.lua]: No, we must use the `set_lru` result: if a `lru_callback` has been set, data might be transformed, so the result is not necessarily the same as the `data` parameter.

https://github.com/thibaultcha/lua-resty-mlcache/pull/29/files/84bdb540283d2f3f439c7fdaa0b75b29ccc2838b#diff-1d36902a4b814b8c38955c37ed51f2c4R280

**Comment by jdesgats** [comment]
Thanks for the quick and thorough review :) It makes a lot of sense. Actually I was using `assert` everywhere in my callbacks :p

TODO list:
- [ ] Rename `lru_callback` to `l1_serializer`
- [ ] Add l1_serializer at the get() level (and keeping it as an instantiation argument too)
- [ ] Add `nil, err` support for `l1_serializer`

File: lib/resty/mlcache.lua (hunk 1 above)
Comment (thibaultcha) [reply lib/resty/mlcache.lua]: Right. Forgive me for I have not touched this library in a while... It is a bit blurry :) I was trying to avoid the extra branch and potential future programming errors now that we can modify `data` on the fly, and need to re-evaluate `== CACHE_MISS_SENTINEL_LRU`. I was hoping maybe we could avoid it, but it's fine as is for now at least 👍

**Comment by Tieske** [comment]
@thibaultcha 

>> 1. I don't like pcall, even with LuaJIT, I've seen the dreaded "yield across C boundary" error. And by then there usually is no way back. So imo better to be clear that the callbacks run unprotected and errors need to be dealt with by the callbacks.
>
>LuaJIT's resumable VM should take care of this?

I thought so as well, but we had those errors somewhere in the global patches (quite a while ago already). Not sure any more how we solved those. Anyway, LuaJIT is less prone to it, but not free of it (opinion based on the errors I referred to, but if you know better, then please disregard my comment)

**Comment by jdesgats** [comment]
OK, I've addressed all your remarks @thibaultcha, not sure how do you want to document the soft error mechanism though: as it is not yet implemented for the cache miss callback, it would feel quite inconsistent.

File: lib/resty/mlcache.lua [hunk 3]
```diff
@@ -513,7 +545,12 @@ function _M:get(key, opts, cb, ...)
 
     -- set our own worker's LRU cache
 
-    set_lru(self, key, data, ttl, neg_ttl)
+    data, err = set_lru(self, key, data, ttl, neg_ttl, l1_serializer)
+    if data == CACHE_MISS_SENTINEL_LRU then
+        data = nil -- convert misses back to nil for return values
+    elseif err then
+        return unlock_and_ret(lock, nil, err)
+    end
```
Comment (thibaultcha) [inline lib/resty/mlcache.lua]: We might save some cycles (and stay consistent) if we take a similar approach as the above L2 hit here:
```lua
data, err = set_lru(self, key, data, ttl, neg_ttl, l1_serializer)
if err then
    return unlock_and_ret(lock, nil, err)
end

if data == CACHE_MISS_SENTINEL_LRU then
    return unlock_and_ret(lock, nil, nil, 2)
end

return unlock_and_ret(lock, data, nil, 2)
```

This way we avoid assigning `data = nil` but keep the same amount of branches in the execution path?

File: t/08-l1_serializer.t [hunk 4]
```diff
@@ -0,0 +1,417 @@
...
+        -- jit.opt.start("hotloop=1")
+        -- jit.opt.start("loopunroll=1000000")
+        -- jit.off()
+    }
+};
+
+run_tests();
+
+__DATA__
+
```
Comment (thibaultcha) [inline t/08-l1_serializer.t]: A first test for each other option in this library is to ensure we validate the argument (in our case, we throw an error if it is not a function). Could we add such a test?

File: t/08-l1_serializer.t [hunk 5]
```diff
@@ -0,0 +1,417 @@
...
+GET /t
+--- response_body
+l1_serializer returned an error: cannot transform
+nil
+--- no_error_log
+[error]
+
+
+
+=== TEST 8: l1_serializer can be given as a :get() parameter
```
Comment (thibaultcha) [inline t/08-l1_serializer.t]: An additional test here might be to assert it is a function again (since it is a different code path in the library)

File: t/08-l1_serializer.t [hunk 6]
```diff
@@ -0,0 +1,417 @@
...
+            ngx.say(value)
+        }
+    }
+--- request
+GET /t
+--- response_body
+true
+transform("value")
+--- no_error_log
+[error]
```
Comment (thibaultcha) [inline t/08-l1_serializer.t]: Maybe we are missing another test: assert that the `l1_serializer` is not called when L2 hit returns a cached miss (upon negative caching, basically). This way we assert our documentation when it says "the argument to `l1_serializer` can never be `nil`"

**Comment by thibaultcha** [comment]
@jdesgats About the soft error return, I am thinking of simply not documenting it quite yet (I know, I know...) but ensuring we implement the same mechanism for the callback very soon (maybe @Tieske can help with that, I remember him volunteering a few weeks ago!) and documenting/releasing both at the same time.

File: t/08-l1_serializer.t (hunk 4 above)
Comment (thibaultcha) [reply t/08-l1_serializer.t]: Another test I just thought of: we need to ensure that we still maintain JIT capabilities when we use `l1_serializer` (granted the later does not use NYI operations of course)

**Comment by jdesgats** [comment]
Sorry I got busy with other things, I think I'll have some time by the end of the week :)

File: README.md [hunk 7]
```diff
@@ -287,6 +292,10 @@ options:
   accepts fractional number parts, like `0.3`. A `neg_ttl` of `0` means the
   cached misses will never expire.
   **Default:** inherited from the instance.
+- `l1_serializer`: function used to transform the values before setting them
+  in the LRU cache (L1). It might be useful to store arbitrary Lua objects
+  such as cdata objects, functions, ... This function is called on L1 miss.
+  **Default:** inherited from the instance.
```
[the remark this answers is not in the document]
Comment (jdesgats) [reply README.md]: I was thinking about a more real-life example: compiling lua source code into a callable function.

I think it would be a better example as it is not something that you could do at any other place, also it exhibits better the error management choices that have to be made (as blindly propagating errors could easily result into error storms).

**Comment by jdesgats** [comment]
Sorry for the long wait, all your further comments are fixed now :)

File: lib/resty/mlcache.lua [hunk 8]
```diff
@@ -269,6 +275,20 @@ local function set_lru(self, key, value, ttl, neg_ttl)
...
+    if l1_serializer then
+        local ok, err
+        ok, value, err = pcall(l1_serializer, value)
+        if not ok then
+            return nil, "l1_serializer threw an error: " .. value
+        elseif value == nil then
+            if err then
+                return nil, "l1_serializer returned an error: " .. err
+            else
+                return nil, "l1_serializer returned a nil value"
```
[the remark this answers is not in the document]
Comment (jdesgats) [reply lib/resty/mlcache.lua]: Done. Actually it is not strictly necessary and wasn't the case initially: it has been added with 0e917465c1e8cc8b6c406ecdc87fa9092e8489da (because is made my troubleshooting more complex at some point, and makes no sense in general I think).

File: README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (thibaultcha) [reply README.md]: Agreed, your example is more appropriate :)

File: t/08-l1_serializer.t [hunk 9]
```diff
@@ -0,0 +1,601 @@
...
+                    return string.format("transform(%q)", s)
+                end
+            }, "value")
+
+            if not ok then
+                ngx.say(err)
+            end
+            ngx.say(tostring(called))
+
+            local value, err = cache:get("key", nil, error)
```
Comment (thibaultcha) [inline t/08-l1_serializer.t]: neat trick to put `error()` here

**Comment by thibaultcha** [comment]
@jdesgats Just merged this to master! One modification I made was removing the `l1_serializer returned an error: ` prefix on `nil` + error (trying to avoid the `foo: bar: thing: error` pattern by delegating context-aware error messages to the user and properly bubbling it up instead).

Thank you very much for your contribution, this has just made mlcache even more awesome than I consider it to be 😅 

Before a new release, I am trying to implement3 more minor things:
- return nil + error from L3 callback
- abstract the IPC module so we can use mlcache's `set()`/`delete()` with something that isn't the bundled `ipc` module
- a `purge()` method (something https://github.com/openresty/lua-resty-lrucache/pull/23 would help a lot for, but we can do without it).

This way we'll be able to use it in Kong instead of the patched copy we are currently using. I'll be trying to get this done soon, I have some time on my hands now.

Thanks again!
